# Assignment: Data Wrangling and Exploratory Data Analysis
## Do Q1 and Q2, and one other question.
`! git clone https://www.github.com/DS3001/assignment2`

**Q1.** Open the "tidy_data.pdf" document in the repo, which is a paper called Tidy Data by Hadley Wickham.

  1. Read the abstract. What is this paper about?
  2. Read the introduction. What is the "tidy data standard" intended to accomplish?
  3. Read the intro to section 2. What does this sentence mean: "Like families, tidy datasets are all alike but every messy dataset is messy in its own way." What does this sentence mean: "For a given dataset, it’s usually easy to figure out what are observations and what are variables, but it is surprisingly difficult to precisely define variables and observations in general."
  4. Read Section 2.2. How does Wickham define values, variables, and observations?
  5. How is "Tidy Data" defined in section 2.3?
  6. Read the intro to Section 3 and Section 3.1. What are the 5 most common problems with messy datasets? Why are the data in Table 4 messy? What is "melting" a dataset?
  7. Why, specifically, is table 11 messy but table 12 tidy and "molten"?
  8. Read Section 6. What is the "chicken-and-egg" problem with focusing on tidy data? What does Wickham hope happens in the future with further work on the subject of data wrangling?

1. The paper is about going over on how to important it is to tidy up data and few people talk about how to tidy up data. This paper will talk about how to tidy up paper with using a small set of tool to do lots of data cleaning. It uses a frame work where it has a specific structure. Each variable is a column, each observation is a row, and each type of observational unit is a table. This makes it easy to model and look at the data.

2.Tidy data standard main goal is to make data cleaning easier and use the same process each time you obtain new data. It stops you from spending a lot of time cleaning data each time and instead use that time to anyazle data.

3. The first sentence means that all tidy datasets have the same identical structure such as having columns with variables, rows with obervations etc. While messy datasets are
different with unique problems such as missing variables or missing rows as well as different formats thats hard to use tools on. The second sentence means that it is easy to
look at observations such as two columns height and legth being clear, but in more complicated cases, its harder to name the observations and variables as it might not be
clear for other people in different environments.

4.A variable: contains all values that measure the same underlying attributes  across units. (like height, temperature, duration)
Observation:contains all values measured on the same unit across attributes. (like a person, or a day, or a race)
values: Either numbers (if quantitative) or strings (if qualitative) of a dataset

5.Tidy data follows a structural with its dataset. A dataset is considered tidy depending on how rows, columns and tables are matched up with observations, variables and types. Tidy data have these characteristics listed.
1. Each variable forms a column.
2. Each observation forms a row.
3. Each type of observational unit forms a table.

6. 5 common problems with messy data is that column headers are values instead of variable names. Multiple variables are stored in one column. Variables are stored in both rows and columns. Multiple types of observational units are stored in the same table. A single observational unit is stored in multiple tables.
 Table 4 is considered messy because it has income spread out across multiple columns instead of using just 1 column income for the data. Melting the data means you turn columns into rows and stack the data into one big column. In this case it would be turning all the income intervals columns and combine then to just one income column that is really long.

7. Table 11 is messy because there are some variables in both rows and columns. The dates are spread out with a lot of unnecessary taking up space with a lot of empty lines. The element column adds on to being messy due to containing variables as it should be its own column. Table 12 is Tidy and Molten because it takes all the all dates columns and combines them to just one column called Date saving up a ton of space. It is finally tidy when it takes the Element column and gives those variables into their own columns. This saves a lot of space.  

8. The chicken-and-egg in tidy data problem refers to either to clean the data first before looking at the data or to look at the data first then clean as you move on.

He hopes that this framework has better data storage strategies and better tools to analyze data. He hopes to have parsing dates, identifying missing values, correcting character encodings where it is all easier to tidy up the data.

**Q2.** This question provides some practice cleaning variables which have common problems.
1. Numeric variable: For `./data/airbnb_hw.csv`, clean the `Price` variable as well as you can, and explain the choices you make. How many missing values do you end up with? (Hint: What happens to the formatting when a price goes over 999 dollars, say from 675 to 1,112?)
2. Categorical variable: For the `./data/sharks.csv` data covered in the lecture, clean the "Type" variable as well as you can, and explain the choices you make.
3. Dummy variable: For the pretrial data covered in the lecture, clean the `WhetherDefendantWasReleasedPretrial` variable as well as you can, and, in particular, replace missing values with `np.nan`.
4. Missing values, not at random: For the pretrial data covered in the lecture, clean the `ImposedSentenceAllChargeInContactEvent` variable as well as you can, and explain the choices you make. (Hint: Look at the `SentenceTypeAllChargesAtConvictionInContactEvent` variable.)

In [200]:
import numpy as np # Import the numpy package into your workspace
import pandas as pd  # Import the pandas package into your workspace

##########################################################################################
###########################################################################################
#Q1
url = "https://raw.githubusercontent.com/uvavmc7ht/assignment2/main/data/airbnb_hw.csv"
df_airbnb = pd.read_csv(url,low_memory=False)

var = 'Price'
#print(df_airbnb[var])
#print(df_airbnb[var].unique(),'\n') # 'n' is not listed in the codebook
#print(df_airbnb[var].value_counts(), '\n') #

df_airbnb[var] = df_airbnb[var].str.replace(",","")
df_airbnb[var] = pd.to_numeric(df_airbnb[var], errors='coerce') # Coerce the variable to numeric

#Replace the comma to empty as we cannot do work with comma numbers. Also made any data that is missing a price as Nan if it is missing so
#it can still work on data.  0 or 1 is wrong as it would mess up math values so nan is better.

##########################################################################################
##########################################################################################
##########################################################################################
##########################################################################################
#Q2 Categorical variable: For the ./data/sharks.csv data covered in the lecture,
#clean the "Type" variable as well as you can, and explain the choices you make.
#Replaced empty cells with 'Nan' to create a dummy variables for missing data.
url_shark = "https://raw.githubusercontent.com/uvavmc7ht/assignment2/main/data/sharks.csv"
df_shark= pd.read_csv(url_shark,low_memory=False)
var = 'Type'

#print(df_shark[var].unique(), "\n")
#print(df_shark[var].value_counts(), '\n')
df_shark[var] = df_shark[var].replace('', np.nan) #any missing data become nan.
df_shark[var] = df_shark[var].replace("Sea Disaster", "Unprovoked") #unlucky sea attack
df_shark[var] = df_shark[var].replace("Watercraft", "Provoked") #change to one category as it is the same provoked attacked on a raft
df_shark[var] = df_shark[var].replace("Boating", "Provoked") #change to one category as it is the same provoked attacked
df_shark[var] = df_shark[var].replace("Boat", "Provoked") #change to one category as it is the same provoked attacked
df_shark[var] = df_shark[var].replace("Boatomg", "Provoked") #change to one category as it is the same provoked attacked on a wooden structures

df_shark[var] = df_shark[var].replace("Questionable", "Unverified") #Same thing
df_shark[var] = df_shark[var].replace("Unconfirmed", "Unverified") #Same meaning
df_shark[var] = df_shark[var].replace("Under investigation", "Unverified") # same meaning
df_shark[var] = df_shark[var].replace("Invalid", "Unverified") # same meaning

#print(df_shark[var].unique(), "\n")
#print(df_shark[var].value_counts(), '\n')




Unprovoked    4955
Provoked       937
Unverified     565
Name: Type, dtype: int64 



In [ ]:
#Q3 Dummy variable: For the pretrial data covered in the lecture, clean the
#WhetherDefendantWasReleasedPretrial variable as well as you can, and, in particular, replace missing values with np.nan.

In [201]:
url = 'http://www.vcsc.virginia.gov/pretrialdataproject/October%202017%20Cohort_Virginia%20Pretrial%20Data%20Project_Deidentified%20FINAL%20Update_10272021.csv'
df = pd.read_csv(url,low_memory=False) # Pandas downloads and loads the .csv file for you

In [169]:
backup_copy = df

In [179]:
df = backup_copy

In [202]:
var = 'WhetherDefendantWasReleasedPretrial'

print(df[var].unique(), "\n")
print(df[var].value_counts(), '\n')

#replace 9 or empty into nan as 1 & 0 are already taken up
df[var] = df[var].replace(9, np.nan)
df[var] = df[var].replace('', np.nan)
print(df[var].unique(), "\n")
print(df[var].value_counts(), '\n')


[9 0 1] 

1    19154
0     3801
9       31
Name: WhetherDefendantWasReleasedPretrial, dtype: int64 

[nan  0.  1.] 

1.0    19154
0.0     3801
Name: WhetherDefendantWasReleasedPretrial, dtype: int64 



In [215]:
##########################################################################################
##########################################################################################
##########################################################################################
##########################################################################################
#Q4 Missing values, not at random: For the pretrial data covered in the lecture, clean the
#ImposedSentenceAllChargeInContactEvent variable as well as you can, and explain the choices
#you make. (Hint: Look at the SentenceTypeAllChargesAtConvictionInContactEvent variable.)

var = 'SentenceTypeAllChargesAtConvictionInContactEvent'

print(df[var].unique(), "\n")


var = 'ImposedSentenceAllChargeInContactEvent'

#print(df[var].unique(), "\n")
df.loc[df['SentenceTypeAllChargesAtConvictionInContactEvent'] == 9,var ] = np.nan # missing
df[var] = df[var].replace(' ', np.nan)  # missing
df.loc[df['SentenceTypeAllChargesAtConvictionInContactEvent'] == 0,var ] = 0 # time should be zero
df.loc[df['SentenceTypeAllChargesAtConvictionInContactEvent'] == 4,var ] = 0 # time should be zero
#print(df[var].unique(), "\n")


[9 0 1 4 2] 

[nan 0 '12' '.985626283367556' '36' '24' '120' '60' '72'
 '11.9917864476386' '84' '180' '4' '96' '2' '44' '5' '115' '132'
 '.328542094455852' '48' '258' '34' '76' '.131416837782341' '111' '9' '3'
 '6' '1.97125256673511' '36.9856262833676' '.0657084188911704'
 '35.4928131416838' '106.492813141684' '8' '54' '18.3141683778234' '480'
 '32' '93' '234' '732' '1.16427104722793' '240' '4.6570841889117' '21' '7'
 '4.49281314168378' '18' '600' '43.1642710472279' '179' '300' '52' '20'
 '192' '702' '14' '55' '5.91375770020534' '30' '108' '68'
 '.657084188911704' '46.6242299794661' '65' '200' '2.95687885010267'
 '24.3285420944559' '12.1642710472279' '117' '81.4928131416838'
 '22.4928131416838' '1980' '3.6570841889117' '56' '10' '2.79260780287474'
 '1' '47' '22' '102' '1500' '40' '284' '11' '42' '162' '156'
 '47.2956878850103' '105' '51' '246' '75' '324' '360' '34.4804928131417'
 '120.328542094456' '66' '59.9917864476386' '29' '660' '51.1642710472279'
 '14.9568788501027' '78' '228' '1.

**Q3.** This question provides some practice doing exploratory data analysis and visualization.

The "relevant" variables for this question are:
  - `level` - Level of institution (4-year, 2-year)
  - `aid_value` - The average amount of student aid going to undergraduate recipients
  - `control` - Public, Private not-for-profit, Private for-profit
  - `grad_100_value` - percentage of first-time, full-time, degree-seeking undergraduates who complete a degree or certificate program within 100 percent of expected time (bachelor's-seeking group at 4-year institutions)

1. Load the `./data/college_completion.csv` data with Pandas.
2. What are are the dimensions of the data? How many observations are there? What are the variables included? Use `.head()` to examine the first few rows of data.
3. Cross tabulate `control` and `level`. Describe the patterns you see.
4. For `grad_100_value`, create a histogram, kernel density plot, boxplot, and statistical description.
5. For `grad_100_value`, create a grouped kernel density plot by `control` and by `level`. Describe what you see. Use `groupby` and `.describe` to make grouped calculations of statistical descriptions of `grad_100_value` by `level` and `control`. Which institutions appear to have the best graduation rates?
6. Create a new variable, `df['levelXcontrol']=df['level']+', '+df['control']` that interacts level and control. Make a grouped kernel density plot. Which institutions appear to have the best graduation rates?
7. Make a kernel density plot of `aid_value`. Notice that your graph is "bi-modal", having two little peaks that represent locally most common values. Now group your graph by `level` and `control`. What explains the bi-modal nature of the graph? Use `groupby` and `.describe` to make grouped calculations of statistical descriptions of `aid_value` by `level` and `control`.
8. Make a scatterplot of `grad_100_value` by `aid_value`. Describe what you see. Now make the same plot, grouping by `level` and then `control`. Describe what you see. For which kinds of institutions does aid seem to increase graduation rates?

**Q4.** This question uses the Airbnb data to practice making visualizations.

  1. Load the `./data/airbnb_hw.csv` data with Pandas. You should have cleaned the `Price` variable in question 2, and you'll need it later for this question.
  2. What are are the dimensions of the data? How many observations are there? What are the variables included? Use `.head()` to examine the first few rows of data.
  3. Cross tabulate `Room Type` and `Property Type`. What patterns do you see in what kinds of rentals are available? For which kinds of properties are private rooms more common than renting the entire property?
  4. For `Price`, make a histogram, kernel density, box plot, and a statistical description of the variable. Are the data badly scaled? Are there many outliers? Use `log` to transform price into a new variable, `price_log`, and take these steps again.
  5. Make a scatterplot of `price_log` and `Beds`. Describe what you see. Use `.groupby()` to compute a desciption of `Price` conditional on/grouped by the number of beds. Describe any patterns you see in the average price and standard deviation in prices.
  6. Make a scatterplot of `price_log` and `Beds`, but color the graph by `Room Type` and `Property Type`. What patterns do you see? Compute a description of `Price` conditional on `Room Type` and `Property Type`. Which Room Type and Property Type have the highest prices on average? Which have the highest standard deviation? Does the mean or median appear to be a more reliable estimate of central tendency, and explain why?
  7. We've looked a bit at this `price_log` and `Beds` scatterplot. Use seaborn to make a `jointplot` with `kind=hex`. Where are the data actually distributed? How does it affect the way you think about the plots in 5 and 6?

**Q5.** Many important datasets contain a race variable, typically limited to a handful of values often including Black, White, Asian, Latino, and Indigenous. This question looks at data gathering efforts on this variable by the U.S. Federal government.

1. How did the most recent US Census gather data on race?
2. Why do we gather these data? What role do these kinds of data play in politics and society? Why does data quality matter?
3. Please provide a constructive criticism of how the Census was conducted: What was done well? What do you think was missing? How should future large scale surveys be adjusted to best reflect the diversity of the population? Could some of the Census' good practices be adopted more widely to gather richer and more useful data?
4. How did the Census gather data on sex and gender? Please provide a similar constructive criticism of their practices.
5. When it comes to cleaning data, what concerns do you have about protected characteristics like sex, gender, sexual identity, or race? What challenges can you imagine arising when there are missing values? What good or bad practices might people adopt, and why?
6. Suppose someone invented an algorithm to impute values for protected characteristics like race, gender, sex, or sexuality. What kinds of concerns would you have?

1.They gathered their data by sending a questionnaire by mail to each household that ask them their race. It would be up to the household people to answer their race and ethnicity.

2.They gather data to get an overview on demographics and to address and race and ethnicity issues that may arise. This plays a major role in politics and society as some politics may either use this data for good or bad. On whether to gain political votes by gerrymandering or during office to determine funding. In society, it helps to see if your group is being represented. Data quality matters, as if error or wrong data can impact on politics address to certain issues. If they see no issue with wrong data, it can be costly and harm certain people or society.


3. They did a good job at how they collected the data. The race category are pretty clear. It is easy to understand and support non English speaker to fill it out by offering it in other languages. The unanimous also helps to get people to fill it out. As well as providing an online version to fill out. One thing they might miss is gathering more data frequently as it can be outdated (like a financial crisis that causes people to move). Large scale survey would be good if they can have like a certain day(s) where people go to a location to fill out the form if they want and this also brings more public awareness. They could add more races or even a blank box if it is too many options. They could get a more accurate data with this Information.

4.They only had two options for sex/gender which was male or female. This can be bad as some people are not neither and it can skew the data as they have to pick one. A better option is to offer more choices or a blank box to consider more people and make the data more accurate and rich.

5. When it comes to data cleaning, one should be careful in not being bias when it comes to sex, gender, sexual identity, or race. Making sure data is fair and non discriminatory. Issues with missing values add challenges as they cannot assume certain parts like gender. Some good practice in cleaning data is being really safe and generlize certain things like gender missing data and put them as nan. However a malicious data cleaner could manipulate the data and put it where it benefits certain groups or attack a group of people.

6. I would wonder how accurate their algorithm is and how they can prove it. I would also wonder if they followed the code of ethics. I would ask if the Algorithm is biased or not and ask them to prove it.

**Q6.** Open the `./data/CBO_data.pdf` file. This contains tax data for 2019, explaining where the money comes from that the U.S. Federal Government Spends in terms of taxation on individuals/families and payroll taxes (the amount that your employer pays in taxes on your wages).

For some context, the Federal government ultimately spent about $4.4 trillion in 2019, which was 21% of GDP (the total monetary value of all goods and services produced within the United States). Individual Income Taxes is the amount individuals pay on their wages to the Federal government, Corporate Income Taxes is the taxes individuals pay on capital gains from investment when they sell stock or other financial instruments, Payroll Taxes is the tax your employer pays on your wages, Excises and Customs Duties are taxes on goods or services like sin taxes on cigarettes or alcohol, and Estate and Gift Taxes are taxes paid on transfers of wealth to other people.

1. Get the Millions of Families and Billions of Dollars data into a .csv file and load it with Pandas.
2. Create a bar plot of individual income taxes by income decile. Explain what the graph shows. Why are some values negative?
3. Create a bar plot of Total Federal Taxes by income decile. Which deciles are paying net positive amounts, and which are paying net negative amounts?
4. Create a stacked bar plot for which Total Federal Taxes is grouped by Individual Income Taxes, Payroll Taxes, Excises and Customs Duties, and Estate and Gift Taxes. How does the share of taxes paid vary across the adjusted income deciles? (Hint: Are these the kind of data you want to melt?)
5. Below the Total line for Millions of Families and Billions of Dollars, there are data for the richest of the richest families. Plot this alongside the bars for the deciles above the Total line. Describe your results.
6. Get the Percent Distribution data into a .csv file and load it with Pandas. Create a bar graph of Total Federal Taxes by income decile.
7. A tax system is progressive if higher-income and wealthier individuals pay more than lower-income and less wealthy individuals, and it is regressive if the opposite is true. Is the U.S. tax system progressive in terms of amount paid? In terms of the percentage of the overall total?
8. Do the rich pay enough in taxes? Defend your answer.